### PR TITLE
fix(core): Better range check for Uni_IsValid() 🙀

### PR DESCRIPTION
**Pull request by srl295**
- rewrite macros as inlines
- optimized approach

Resolves TODO in #9293

For: #7375

@keymanapp-test-bot skip

### CODE REVIEW

File: core/src/kmx/kmx_xstring.h [hunk 1]
```diff
@@ -149,6 +171,50 @@ u16string_to_u32string(const std::u16string &source) {
...
+inline bool Uni_IsEndOfPlaneNonCharacter(km_kbp_usv ch) {
+  return (((ch) & Uni_FFFE_NONCHARACTER) == Uni_FFFE_NONCHARACTER); // matches FFFF or FFFE
+}
+
+inline bool Uni_IsNoncharacter(km_kbp_usv ch) {
+  return (((ch) >= Uni_FD_NONCHARACTER_START && (ch) <= Uni_FD_NONCHARACTER_END) || Uni_IsEndOfPlaneNonCharacter(ch));
+}
+
+inline bool Uni_InCodespace(km_kbp_usv ch) {
+  return ((ch) <= 0x10FFFF);
```
Comment (mcdurdin) [inline core/src/kmx/kmx_xstring.h:183]: Can we get a const for this as well given we have all the others?

File: core/src/kmx/kmx_xstring.h [hunk 2]
```diff
@@ -149,6 +171,50 @@ u16string_to_u32string(const std::u16string &source) {
...
+}
+
+inline bool Uni_IsValid(km_kbp_usv start, km_kbp_usv end) {
+  // quicker check
+  if (!Uni_IsValid(end) || !Uni_IsValid(start) || (end < start)) {  // start is checked below
+    return false;
+  }
+
+  // If 'start' is low enough in the BMP, we need to avoid (1) Surrogates and (2) the FDxx nonchars
+  if (start < Uni_FD_NONCHARACTER_END) {
```
Comment (mcdurdin) [inline core/src/kmx/kmx_xstring.h:197]: Do we need this first test? The child tests test start anyway?

**Review by mcdurdin — APPROVED** [review]
LGTM, minor nits only

File: core/src/kmx/kmx_xstring.h (hunk 2 above)
Comment (srl295) [reply core/src/kmx/kmx_xstring.h:197]: yes, it was meant as a heuristic but probably fails at it

Comment (srl295) [reply core/src/kmx/kmx_xstring.h]: done

File: core/src/kmx/kmx_xstring.h (hunk 1 above)
Comment (srl295) [reply core/src/kmx/kmx_xstring.h]: done

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 17.0.149-alpha](https://keyman.com/downloads/releases/alpha/17.0.149)

**Pull request merged**